ff/README.md

```diff
@@ -36,3 +36,103 @@ The above two models serve as abstractions for constructing the extension fields
 - [`Fp6_3over2`](https://github.com/arkworks-rs/algebra/blob/master/ff/src/fields/models/fp6_3over2.rs#L49) - Extension tower, similar to the above except that the towering order is reversed: it's a cubic extension on a quadratic extension field, i.e. `BaseField = Fp2`, but `BasePrimeField = Fp`. Only this latter one is exported by default as `Fp6`.
 - [`Fp12_2over3over2`](https://github.com/arkworks-rs/algebra/blob/master/ff/src/fields/models/fp12_2over3over2.rs#L83) - Extension tower: quadratic extension of `Fp6_3over2`, i.e. `BaseField = Fp6`.
+
+## Usage
+
+There are two important traits when working with finite fields: [`Field`],
+and [`PrimeField`]. Let's explore these via examples.
+
+### [`Field`]
+
+The [`Field`] trait provides a generic interface for any finite field.
+Types implementing [`Field`] support common field operations
+such as addition, subtraction, multiplication, and inverses.
+
+```rust
+use ark_ff::Field;
+// We'll use a field associated with the BLS12-381 pairing-friendly
+// group for this example.
+use ark_test_curves::bls12_381::Fq2 as F;
+// `ark-std` is a utility crate that enables `arkworks` libraries
+// to easily support `std` and `no_std` workloads, and also re-exports
+// useful crates that should be common across the entire ecosystem, such as `rand`.
+use ark_std::{One, UniformRand};
+
+let mut rng = ark_std::test_rng();
+// Let's sample uniformly random field elements:
+let a = F::rand(&mut rng);
+let b = F::rand(&mut rng);
+
+// We can add...
+let c = a + b;
+// ... subtract ...
+let d = a - b;
+// ... double elements ...
+assert_eq!(c + d, a.double());
+
+// ... multiply ...
+let e = c * d;
+// ... square elements ...
+assert_eq!(e, a.square() - b.square());
+
+// ... and compute inverses ...
+assert_eq!(a.inverse().unwrap() * a, F::one()); // have to to unwrap, as `a` could be zero.
+```
+
+In some cases, it is important to take square roots of field elements
+(e.g.: for point compression of elliptic curve elements.)
+To support this, users can implement the `sqrt`-related methods for their field type. These can be used as follows:
+
+```rust
+use ark_ff::Field;
+// As before, we'll use a field associated with the BLS12-381 pairing-friendly
+// group for this example.
+use ark_test_curves::bls12_381::Fq2 as F;
+use ark_std::{One, UniformRand};
+
+let mut rng = ark_std::test_rng();
+// Let's try to sample a random square via rejection sampling:
+let mut a = F::rand(&mut rng);
+while a.legendre().is_qnr() { // A square is also called a *quadratic residue*
+    a = F::rand(&mut rng);
+}
+
+// Since `a` is a square, we can compute its square root:
+let b = a.sqrt().unwrap();
+assert_eq!(b.square(), a);
+
+// Let's sample a random *non-square*
+let mut a = F::rand(&mut rng);
+while a.legendre().is_qr() {
+    a = F::rand(&mut rng);
+}
+// The square root should not exist:
+assert_eq!(a.sqrt(), None);
```

Comment:
I feel like this section is simpler, if we don't do legendre and is_quadratic_residue, but instead:
* Sample a square directly (just square it ourselves)
* Make it a non-square via multiplying by a known non-square

Reply:
Since it's for illustrative purposes, we could simplify even further and get rid of the loops in favour of a simple if/else:
```
let a = F::rand(&mut rng);

if a.legendre().is_qnr() {
    assert_eq!(a.sqrt(), None);
} else {
    let b = a.sqrt().unwrap();
    assert_eq!(b.square(), a);
}
```

Reply:
Incorporated your suggestion @mmagician.